### ISSUE TITLE
nucleo-matcher: Normalization is incomplete.

### ISSUE THREAD

**Issue by markus-bauer** [helix-editor/nucleo#51]
Thanks for this crate. But I had some issues with normalization.

Here are some random examples from the polish alphabet, that don't work as expected; there might be more issues, though.
```rust
use nucleo_matcher::pattern::{AtomKind, CaseMatching, Normalization};

fn main() {
    // Example: https://en.wikipedia.org/wiki/Polish_alphabet
    let tests = [
        ('ą', 'a'),
        ('ć', 'c'),
        ('ę', 'e'),
        ('ł', 'l'),
        ('ń', 'n'),
        ('ó', 'o'),
        ('ś', 's'),
        ('ź', 'z'),
        ('ż', 'z'),
        ('Ą', 'A'),
        ('Ć', 'C'),
        ('Ę', 'E'),
        ('Ł', 'L'),
        ('Ń', 'N'),
        ('Ó', 'O'),
        ('Ś', 'S'),
        ('Ź', 'Z'),
        ('Ż', 'Z'),
    ];

    // test normalization funcion
    println!("--- normalize ---");
    for (c, expected) in tests {
        let normalized = nucleo_matcher::chars::normalize(c);
        if normalized == expected {
            println!("okay: {} -> {}.", c, normalized);
        } else {
            println!("fail: {} -> {}. expected {}", c, normalized, expected);
            let c_lower = c.to_lowercase().next().unwrap();
            let expected_lower = expected.to_lowercase().next().unwrap();
            let normalized_lower = nucleo_matcher::chars::normalize(c_lower);
            if normalized_lower == expected_lower {
                println!(" but lowercase works: {} -> {}", c_lower, normalized_lower);
            }
        }
    }

    // test matcher
    println!("--- matcher ---");
    let mut matcher = nucleo_matcher::Matcher::new(nucleo_matcher::Config::DEFAULT);

    for (c, expected) in tests {
        // search for c in expected
        let haystack = nucleo_matcher::Utf32String::from(c.to_string());
        let atom = nucleo_matcher::pattern::Atom::new(
            &expected.to_string(),
            CaseMatching::Smart,
            Normalization::Smart,
            AtomKind::Exact,
            false,
        );

        if atom.score(haystack.slice_u32(..), &mut matcher).is_some() {
            println!("okay: {}", c)
        } else {
            println!("fail: {} not in {}", atom.needle_text(), haystack)
        }
    }
}
```
```
--- normalize ---
okay: ą -> a.
okay: ć -> c.
okay: ę -> e.
okay: ł -> l.
okay: ń -> n.
okay: ó -> o.
okay: ś -> s.
okay: ź -> z.
okay: ż -> z.
fail: Ą -> Ą. expected A
 but lowercase works: ą -> a
fail: Ć -> Ć. expected C
 but lowercase works: ć -> c
fail: Ę -> Ę. expected E
 but lowercase works: ę -> e
fail: Ł -> Ł. expected L
 but lowercase works: ł -> l
fail: Ń -> Ń. expected N
 but lowercase works: ń -> n
okay: Ó -> O.
fail: Ś -> Ś. expected S
 but lowercase works: ś -> s
fail: Ź -> Ź. expected Z
 but lowercase works: ź -> z
fail: Ż -> Ż. expected Z
 but lowercase works: ż -> z
--- matcher ---
okay: ą
okay: ć
okay: ę
okay: ł
okay: ń
okay: ó
okay: ś
okay: ź
okay: ż
fail: A not in Ą
fail: C not in Ć
fail: E not in Ę
fail: L not in Ł
fail: N not in Ń
okay: Ó
fail: S not in Ś
fail: Z not in Ź
fail: Z not in Ż
```

Also, I think you might have made some copy-paste/codegen mistakes (here and in other places):
https://github.com/helix-editor/nucleo/blob/6df3cd0b39cd83e3c176ac4528136fcdd8c1f79a/matcher/src/chars/normalize.rs#L340-L390

**Comment by alexrutar** [comment]
I might pick this up at some point; but if some else does (and for my own reference in the future) I have linked the Wikipedia pages for reference to the usual Unicode blocks:

- https://en.wikipedia.org/wiki/Latin-1_Supplement
- https://en.wikipedia.org/wiki/Latin_Extended_Additional
- https://en.wikipedia.org/wiki/Latin_Extended-A
- https://en.wikipedia.org/wiki/Latin_Extended-B

And then there are some weirder ones (not sure if these should be handled, and if so which planes)

- https://en.wikipedia.org/wiki/Latin_Extended-C
- https://en.wikipedia.org/wiki/Latin_Extended-D
- https://en.wikipedia.org/wiki/Superscripts_and_Subscripts
- https://en.wikipedia.org/wiki/IPA_Extensions
- https://en.wikipedia.org/wiki/Phonetic_Extensions
- https://en.wikipedia.org/wiki/Phonetic_Extensions_Supplement
- https://en.wikipedia.org/wiki/Halfwidth_and_Fullwidth_Forms_(Unicode_block)
- https://en.wikipedia.org/wiki/Letterlike_Symbols

Not sure if I am missing anything else. For codegen the [unicode-normalization](https://docs.rs/unicode-normalization/latest/unicode_normalization/) crate should work. I guess hand-rolling an optimized normalizer for latin-only blocks is quite good for performance reasons, especially because of the way graphemes are handled internally (i.e. we do not need to handle multi-char normalization since we just grab the first char in each grapheme cluster anyway).

There's also [icu_normalizer](https://docs.rs/icu_normalizer/latest/icu_normalizer/) for comparison. I guess any implementation should be benchmarked against `icu_normalizer` since I think it is the best-performing fully featured normalizer implementing all of [UAX 15](https://www.unicode.org/reports/tr15/) which is also written in Rust. Though a custom implementation should be substantially faster / smaller.

**Comment by pascalkuthe** [comment]
This likely just  a bug with the current lookup but the implementation is indeed optimized for performance and I don't want something exhaustive like the existing creates you linked here.

The tabels are copied from fzf and the normalization option in nucleo and fzf are specifically only about latin characters.

**Comment by alexrutar** [comment]
This is now fixed in #60
